File: algorithms/096.Unique_Binary_Search_Trees/Unique_Binary_Search_Trees.cpp
/*  Descripton:  Interval dp
               dp[i] = dp[k] * dp[i-k-1] { 0<=k<=i-1 }
*/

#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int numTrees(int n) {
        vector<int> dp(n + 1);
        dp[1] = dp[0] = 1;
        for (int i = 2; i <= n; i++) {
            for (int k = 0; k < i; k++)
                dp[i] += dp[k] * dp[i - k - 1];
        }
        
        return dp[n];
    }
};

int main() {

    return 0;
}

File: algorithms/096.Unique_Binary_Search_Trees/Unique_Binary_Search_Trees.py
<<<<<<< HEAD
# Time:  O(n^2)
# Space: O(n)
class Solution:
    # @return an integer
    def numTrees(self, n):
        counts = [1, 1]
        for i in xrange(2, n + 1):
            count = 0
            for j in xrange(i):
                count += counts[j] * counts[i - j - 1]
            counts.append(count)
        return counts[-1]
    
if __name__ == "__main__":
    print Solution().numTrees(3)
=======
# Time:  O(n^2)
# Space: O(n)
class Solution:
    # @return an integer
    def numTrees(self, n):
        counts = [1, 1]
        for i in xrange(2, n + 1):
            count = 0
            for j in xrange(i):
                count += counts[j] * counts[i - j - 1]
            counts.append(count)
        return counts[-1]
    
if __name__ == "__main__":
    print Solution().numTrees(3)
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
